platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/EmbeddedSolrFactory.java

```diff
@@ -46,167 +42,176 @@
-      SolrConfig solrConfig =
-          new SolrConfig(
-              Paths.get(solrConfigHome.getParent()),
-              solrConfigFileName,
-              new InputSource(FileUtils.openInputStream(solrConfigFile)));
-      IndexSchema indexSchema =
-          new IndexSchema(
-              solrConfig,
-              schemaFileName,
-              new InputSource(FileUtils.openInputStream(solrSchemaFile)));
-      SolrResourceLoader loader =
-          new SolrResourceLoader(Paths.get(solrConfigHome.getAbsolutePath()));
-      SolrCoreContainer container = new SolrCoreContainer(loader);
-
-      String dataDirPath = null;
-      File dataDir = configProxy.getDataDirectory();
-      if (dataDir != null) {
-        dataDirPath = Paths.get(dataDir.getAbsolutePath(), coreName, "data").toString();
-        LOGGER.debug("Using data directory [{}]", dataDirPath);
-      }
-      CoreDescriptor coreDescriptor =
-          new CoreDescriptor(
+      Thread.currentThread().setContextClassLoader(EmbeddedSolrFactory.class.getClassLoader());
+      final SolrResourceLoader loader = closer.with(newLoader(files.getConfigHome().toPath()));
+      final SolrCoreContainer container = newContainer(loader);
+      final SolrConfig solrConfig = files.getConfig();
+      final CoreDescriptor coreDescriptor =
+          newDescriptor(
               coreName, solrConfig.getResourceLoader().getInstancePath(), new Properties(), false);
+      final IndexSchema indexSchema =
+          EmbeddedSolrFactory.INDEX_CACHE.computeIfAbsent(files, EmbeddedSolrFiles::getSchemaIndex);
+      final SolrCore core =
+          closer.with(
+              newCore(
+                  container,
+                  coreName,
+                  files.getDataDirPath(),
+                  solrConfig,
+                  indexSchema,
+                  null,
+                  coreDescriptor,
+                  null,
+                  null,
+                  null,
+                  false));
 
-      SolrCore core =
-          new SolrCore(
-              container,
-              coreName,
-              dataDirPath,
-              solrConfig,
-              indexSchema,
-              null,
-              coreDescriptor,
-              null,
-              null,
-              null,
-              false);
       container.register(coreName, core, false, true);
-
-      return new EmbeddedSolrServer(container, coreName);
-    } catch (ParserConfigurationException | IOException | SAXException e) {
-      throw new IllegalArgumentException(
-          "Unable to parse Solr configuration file: " + solrConfigFileName, e);
+      return closer.returning(newServer(container, coreName));
     } finally {
       Thread.currentThread().setContextClassLoader(tccl);
     }
   }
 
-  public static File getConfigFile(
-      String configFileName, ConfigurationFileProxy configProxy, String core) {
-    return FileUtils.toFile(configProxy.getResource(configFileName, core));
+  @VisibleForTesting
```

Comment:
❓ This class is a pain to test ☹️ I wonder what's better here. Make all of those methods `@VisibleForTesting` and have to extend the class to override them, or inject `Supplier<>` instances and mock those out.

Reply:
I hate this class like you have no idea. I started trying to re-work it to make it easier to test and the scope of that change grew way beyond what I felt was acceptable. The biggest issue is that there is no way to mock the Solr classes constructor or what is inside so we are forced to create factory methods like that and intercept those.

This is where I missed the ability to do mocking of static functions or constructors when it comes to 3rdParty stuff.